decidim/decidim
# Fix data consent expiry

#### :tophat: What? Why?
This fixes the issue described at #9569 that the data consent is not stored across different browser sessions. After closing the browser and going back to the site, the user needs to re-accept the cookies whereas in the past the consent persisted for 365 days.

This adds the expiry time to the data consent cookie and also addresses few other issues regarding this functionality as agreed before:
- Also set the `SameSite` and `Secure` cookies according to the site's configuration
- Rename the `cookie_consent` strings to `data_consent` and `cookie_details` strings to just `details` not to confuse implementors that the data consent is not only about cookies (this change was agreed today 2022-07-13 in the maintainers meeting)

I am also introducing the ability to test site under HTTPS in both tests and the development app in order to test the `secure` flag for the cookie correctly.

#### :pushpin: Related Issues
- Fixes #9569

#### Testing
To test the original issue:
- Remove the development app's assets to let Webpacker recompile them for you
- Start the rails server
- Go to the site
- Accept cookies/give data consent
- Close all browser windows (even those in private browsing mode)
- Re-open the browser
- Go back to the site to see that you don't see the cookie consent dialog anymore by default

To test the `secure` flag under HTTPS (and also the `SameSite` flag at the same time):
- Regenerate the development app with this PR applied
- Start the rails server using `DEV_SSL=true bundle exec rails s` (binds the server under HTTPS to port 3443)
- Clear all cookies for the local development domain you are using (e.g. `localhost`)
- Go to `https://localhost:3443`
- Accept cookies/give data consent
- Check the cookie flags from the developer console for the `localhost` domain
- See that the flags are correct

### :camera: Screenshots
![Sample cookies at localhost](https://user-images.githubusercontent.com/864340/178709486-cccbbc2d-1415-4117-92c4-279c23909dc7.png)
